### PR TITLE
feat: added support for custom parser for markdown to document

### LINKED ISSUE #401: [FR] Add Support for custom Parsers in mardownTODocument Function
### Description

Add support to pass custom parsers to markdownToDocument function

### Impact

Anyone who has  a  custom markdown and wants to convert it to  Document can utilise the custom parsers

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Mukund-Tandon**
fixes: #401

### CODE REVIEW

**Comment by LucasXu0** [comment]
@Mukund-Tandon Please format the code

**Comment by Mukund-Tandon** [comment]
> @Mukund-Tandon Please format the code

Ok,I am also writing test for it

File: lib/src/plugins/markdown/decoder/delta_markdown_decoder.dart [hunk 1]
```diff
@@ -56,6 +60,12 @@ class DeltaMarkdownDecoder extends Converter<String, Delta>
       _attributes[BuiltInAttributeKey.href] = element.attributes['href'];
     } else if (element.tag == 'u') {
       _attributes[BuiltInAttributeKey.underline] = true;
+    } else {
+      for (final key in element.attributes.keys) {
+        if (element.attributes[key] != null) {
+          _attributes[key] = jsonDecode(element.attributes[key]!);
+        }
+      }
```
Comment (LucasXu0) [inline lib/src/plugins/markdown/decoder/delta_markdown_decoder.dart:68]:
```suggestion
      element.attributes.forEach((key, value) {
        _attributes[key] = jsonDecode(value);
      });
```

File: lib/src/plugins/markdown/decoder/delta_markdown_decoder.dart [hunk 2]
```diff
@@ -72,6 +82,10 @@ class DeltaMarkdownDecoder extends Converter<String, Delta>
       _attributes.remove(BuiltInAttributeKey.href);
     } else if (element.tag == 'u') {
       _attributes.remove(BuiltInAttributeKey.underline);
+    } else {
+      for (var key in element.attributes.keys) {
```
Comment (LucasXu0) [inline lib/src/plugins/markdown/decoder/delta_markdown_decoder.dart:86]:
```suggestion
      for (final key in element.attributes.keys) {
```

File: lib/src/plugins/markdown/decoder/document_markdown_decoder.dart [hunk 3]
```diff
@@ -26,7 +34,8 @@ class DocumentMarkdownDecoder extends Converter<String, Document> {
       if (lines[i].startsWith('```') &&
           lines[i].endsWith('```') &&
           lines[i].length > 3) {
-        document.insert([i], [_convertLineToNode(lines[i])]);
+        document
+            .insert([i], [_convertLineToNode(lines[i], customInlineSyntaxes)]);
```
Comment (LucasXu0) [inline lib/src/plugins/markdown/decoder/document_markdown_decoder.dart:38]:
```suggestion
        document.insert(
          [i],
          [_convertLineToNode(lines[i], customInlineSyntaxes)],
        );
```

File: lib/src/plugins/markdown/decoder/document_markdown_decoder.dart [hunk 4]
```diff
@@ -40,24 +49,42 @@ class DocumentMarkdownDecoder extends Converter<String, Document> {
...
+          document.insert(
+            [tempLinePointer],
+            [_convertLineToNode(codeBlock, customInlineSyntaxes)],
+          );
           i++;
         }
       } else {
-        document.insert([i], [_convertLineToNode(lines[i])]);
+        document
+            .insert([i], [_convertLineToNode(lines[i], customInlineSyntaxes)]);
```
Comment (LucasXu0) [inline lib/src/plugins/markdown/decoder/document_markdown_decoder.dart:67]:
```suggestion
        document.insert(
          [i],
          [_convertLineToNode(lines[i], customInlineSyntaxes)],
        );
```

File: lib/src/plugins/markdown/decoder/document_markdown_decoder.dart [hunk 5]
```diff
@@ -40,24 +49,42 @@ class DocumentMarkdownDecoder extends Converter<String, Document> {
...
+  ) {
+    final decoder =
+        DeltaMarkdownDecoder(customInlineSyntaxes: customInlineSyntaxes);
+    for (final parser in customNodeParsers) {
+      final Node? node = parser.transform(line);
+
+      if (node != null) {
+        return node;
+      }
+    }
```
Comment (LucasXu0) [inline lib/src/plugins/markdown/decoder/document_markdown_decoder.dart:87]:
```suggestion
    final decoder = DeltaMarkdownDecoder(
      customInlineSyntaxes: customInlineSyntaxes,
    );
    for (final parser in customNodeParsers) {
      final node = parser.transform(line);
      if (node != null) {
        return node;
      }
    }
```

**Comment by LucasXu0** [comment]
@Mukund-Tandon I refactored some codes. The suggestions are just for your reference, don't commit them.

**Comment by Mukund-Tandon** [comment]
> @Mukund-Tandon I refactored some codes. The suggestions are just for your reference, don't commit them.

Ok 👍

**Pull request merged**